src/suit_decoder.c

```diff
@@ -11,68 +11,16 @@
 #include <cose_decode.h>
 #include <suit_manifest.h>
 
-/** Verify that the header matches the expected length.
- */
-static bool header_check(const uint8_t *cbstr, size_t exp_len)
-{
-	uint8_t byte = *cbstr;
-
-	/* Verify the expected length by decoding the ZCBOR header manually.
-	 *
-	 * The first byte indicates the type of the data:
-	 *  - 0x40:0x58 is a byte string with the length encoded as a sum with 0x40.
-	 *  - 0x58 is a byte string with the length encoded on the next byte.
-	 *  - 0x59 is a byte string with the length encoded on the next two bytes.
-	 *  - 0x80:0x88 is an array with the length encoded as a sum with 0x80.
-	 *  - 0x98 is an array with the length encoded on the next byte.
-	 *  - 0x99 is an array with the length encoded on the next two bytes.
-	 *
-	 * Indefinite length structures, as well as those with length encoded on
-	 * 4 bytes (0x5A and 0x9A) are not supported.
-	 * For more details, see the description of major types in RFC 7049.
-	 *
-	 * It is expected, that the decoding will be covered by the ZCBOR library
-	 * in the near future, but currently the implementation of the
-	 * zcbor_header_len(..) is not robust enough to pass the unit tests.
-	 */
-	if ((byte >= 0x40) && (byte < 0x58)) {
-		return (byte == (0x40 + exp_len));
-	} else if (byte == 0x58) {
-		return (cbstr[1] == exp_len);
-	} else if (byte == 0x59) {
-		return ((cbstr[1] * 256 + cbstr[2]) == exp_len);
-	} else if ((byte >= 0x80) && (byte < 0x88)) {
-		return (byte == (0x80 + exp_len));
-	} else if (byte == 0x98) {
-		return (cbstr[1] == exp_len);
-	} else if (byte == 0x99) {
-		return ((cbstr[1] * 256 + cbstr[2]) == exp_len);
-	}
-
-	return false;
-}
-
+/** Extract the major type, i.e. the first 3 bits of the header byte. */
+#define MAJOR_TYPE(header_byte) ((zcbor_major_type_t)(((header_byte) >> 5) & 0x7))
 
 /** Calculate the length of the CBOR byte string and array header
  */
-static int header_len(size_t len, const uint8_t *value)
+static int header_len(size_t len, const uint8_t *value, zcbor_major_type_t major_type)
 {
-	int exp_len = -1;
-	bool header_ok = false;
-
-	if (len <= 23) {
-		exp_len = 1;
-	} else if (len <= 255) {
-		exp_len = 2;
-	} else if (len <= 0xFFFF) {
-		exp_len = 3;
-	}
+	int exp_len = zcbor_header_len(len);
```

Comment:
API not yet available, do you know when the new ZCBOR will make it's way to the `-next` repositories?
Alternatively - an `#ifdef` on the ZCBOR library version may help here.

Reply:
It's been in upstream zephyr since april, so if they do an upmerge you will get it. I also opened a fromtree PR: https://github.com/nrfconnect/sdk-nrf-next/pull/1599